src/libraries/System.Private.CoreLib/src/System/IO/Path.cs

```diff
@@ -926,5 +926,70 @@ private static string GetRelativePath(string relativeTo, string path, StringComp
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A string containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="path"/> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentException"><paramref name="path" /> is empty.</exception>
+        public static string RemoveRedundantSegments(string path)
+        {
+            if (path == null)
+                throw new ArgumentNullException(nameof(path));
+
+            return RemoveRelativeSegments(path.AsSpan());
+        }
+
+        /// <summary>
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A read-only span containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        /// <exception cref="ArgumentException"><paramref name="path" /> is empty.</exception>
+        public static string RemoveRedundantSegments(ReadOnlySpan<char> path)
+        {
+            if (PathInternal.IsEffectivelyEmpty(path))
+                throw new ArgumentException(SR.Arg_PathEmpty, nameof(path));
+
+            return PathInternal.RemoveRelativeSegments(path);
+        }
+
+        /// <summary>
+        /// Tries to remove any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A read-only span containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <param name="destination">When the method returns <see langword="true" />, a span containing <paramref name="path" /> without any redundant segments; when the method returns <see langword="false" />, this parameter remains unmodified.</param>
+        /// <param name="charsWritten">When the method returns <see langword="true" />, contains the total number of characters written in <paramref name="destination" />; when the method returns <see langword="false" />, the value is zero.</param>
+        /// <returns><see langword="true" /> if the operation succeeds; <see langword="false" /> if the operation fails.</returns>
+        /// <remarks>The method returns `false` when the path is empty, when the destination span's length is less than the path's length, or when attempting to remove the relative segments results in a failure.</remarks>
+        public static bool TryRemoveRedundantSegments(ReadOnlySpan<char> path, Span<char> destination, out int charsWritten)
+        {
+            charsWritten = 0;
+
+            if (PathInternal.IsEffectivelyEmpty(path))
+                return false;
+
+            if (destination.Length < path.Length)
```

Comment:
There is nothing wrong with having smaller destination if there is a lot of redundant segments.

Reply:
I meant the opposite. The destination path should not be larger than the original path.

Reply:
What if `path` has zero redundant segments? Then `destination` should have the exact same string. The length of `destination` should be exactly the same as `path`.

Reply:
This method signature pattern is used for number of APIs. One example from many:

https://docs.microsoft.com/en-us/dotnet/api/system.buffers.text.utf8formatter.tryformat?view=netcore-3.1#System_Buffers_Text_Utf8Formatter_TryFormat_System_Double_System_Span_System_Byte__System_Int32__System_Buffers_StandardFormat_

The contract implemented by these API should be consistent, i.e.:
- `charsWritten`: When the method returns, contains number of characters written into destination buffer.
- Return: `true` if the operation succeeds; `false` if buffer is too small.